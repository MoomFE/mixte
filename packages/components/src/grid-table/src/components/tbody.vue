<!-- 表体 -->

<template>
  <div class="mixte-gt-tbody">
    <template v-for="(record, index) in props.data" :key="record[rowKey]">
      <!-- 表格行 -->
      <div class="mixte-gt-tr">
        <!-- 表体单元格 -->
        <template v-for="column in props.columns" :key="column.field">
          <Td :column :record :index />
        </template>
      </div>
    </template>
  </div>
</template>

<script lang="ts" setup>
  import { useShared } from '../composables/useShared';
  import Td from './td.vue';

  const { props, rowKey } = useShared()!;
</script>
