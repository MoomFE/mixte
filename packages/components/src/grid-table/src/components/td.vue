<template>
  <div class="mixte-gt-cell mixte-gt-td" :class="[classes, cellClasses]" :style>
    <!-- 展开子级数据的按钮 -->
    <button
      v-if="isExpandVisible"
      class="mixte-gt-cell-expand-icon"
      :class="{ 'mixte-gt-cell-expand-icon-spaced': expandIconSpaced }"
      :style="{ marginLeft: `${(node?.level ?? 0) * 15}px` }"
      @click="updateExpanded(record[rowKey])"
    >
      <template v-if="!expandIconSpaced">
        <i-ant-design-minus-outlined v-if="expandedRowSet.has(node.key as string)" />
        <i-ant-design-plus-outlined v-else />
      </template>
    </button>

    <!-- 自定义渲染 -->
    <component
      v-if="isFunction(column.render)"
      :is="column.render"
      :value :record :column :index
    />
    <!-- 指定字段单元格插槽 -->
    <slot
      v-else-if="$slots[`cell-${column.field}`]"
      :name="`cell-${column.field}`"
      :value :record :column :index
    />
    <!-- 通用字段单元格插槽 -->
    <slot
      v-else-if="$slots.cell"
      name="cell"
      :value :record :column :index
    />
    <!-- 值 -->
    <template v-else>
      {{ value }}
    </template>
  </div>
</template>

<script lang="ts" setup>
  import type { GridTableColumn, GridTableFieldsSlots } from '@mixte/components/grid-table/types';
  import type { TreeNode } from 'treemate';
  import { isFunction } from 'mixte';
  import { computed } from 'vue';
  import { useCell } from '../composables/useCell';
  import { useShared } from '../composables/useShared';

  interface Props {
    node: TreeNode<any>;
    column: GridTableColumn<Record<string, any>>;
    index: number;
  }

  const props = defineProps<Props>();

  defineSlots<GridTableFieldsSlots<any>>();

  const { expandedRowSet, rowKey, childrenKey, updateExpanded } = useShared()!;
  const { createCellStore } = useCell()!;
  const { classes, cellClasses, style, isExpandVisible } = createCellStore(props.column.field, props.column);

  const record = computed(() => props.node.rawNode);
  const value = computed(() => record.value[props.column.field]);

  const expandIconSpaced = computed(() => !record.value[childrenKey.value]?.length);
</script>
