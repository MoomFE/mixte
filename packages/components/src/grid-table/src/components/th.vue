<template>
  <div
    ref="thRef"
    class="mixte-gt-cell mixte-gt-th"
    :class="[cellClasses, column.cellClass, column.headerCellClass]"
    :style="[cellStyle, { zIndex: (zIndex ?? 0) + 1 }]"
    :data-field="column.field"
  >
    <!-- 自定义渲染 -->
    <component v-if="isFunction(column.headerRender)" :is="column.headerRender" :column />
    <!-- 指定字段表头单元格插槽 -->
    <slot v-else-if="$slots[`header-${column.field}`]" :name="`header-${column.field}`" :column />
    <!-- 通用表头单元格插槽 -->
    <slot v-else-if="$slots.header" name="header" :column />
    <!-- 正常渲染 -->
    <template v-else> {{ column.title }}</template>
  </div>
</template>

<script lang="ts" setup>
  import type { GridTableColumn, GridTableHeaderSlots } from '@mixte/components/grid-table/types';
  import { isFunction } from 'mixte';
  import { useCell, useTh } from '../composables/useCell';

  interface Props {
    column: GridTableColumn<Record<string, any>>;
  }

  const props = defineProps<Props>();

  defineSlots<GridTableHeaderSlots<any>>();

  const { createColumnStore } = useCell()!;
  const { cellClasses, cellStyle, zIndex } = createColumnStore(props.column.field, props.column);

  const { thRef } = useTh(props.column);
</script>
