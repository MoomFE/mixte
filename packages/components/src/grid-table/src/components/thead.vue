<!-- 表头 -->

<template>
  <div ref="tableTheadRef" class="mixte-gt-thead">
    <!-- 表格行 -->
    <div class="mixte-gt-tr">
      <!-- 表头单元格 -->
      <template v-for="column in props.columns" :key="column.field">
        <component :is="h(Th, { column }, $slots)" />
      </template>
    </div>
  </div>
</template>

<script lang="ts" setup>
  import type { GridTableHeaderSlots } from '@mixte/components/grid-table/types';
  import { useShared } from '../composables/useShared';
  import Th from './th.vue';

  defineSlots<GridTableHeaderSlots<any>>();

  const {
    props,
    tableTheadRef,
  } = useShared()!;
</script>
