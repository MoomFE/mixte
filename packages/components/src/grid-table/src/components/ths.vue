<!-- 整行表头单元格 -->

<template>
  <template v-for="column in columns" :key="column.field">
    <component :is="h(Th, { column }, $slots)" />
  </template>
</template>

<script lang="ts" setup>
  import type { GridTableHeaderSlots } from '@mixte/components/grid-table/types';
  import { h } from 'vue';
  import { useShared } from '../composables/useShared';
  import Th from './th.vue';

  defineSlots<GridTableHeaderSlots<any>>();

  const {
    columns,
  } = useShared()!;
</script>
