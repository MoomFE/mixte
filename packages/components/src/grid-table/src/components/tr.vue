<template>
  <div ref="trRef" class="mixte-gt-tr" :data-index="index">
    <template v-for="column in tableProps.columns" :key="column.field">
      <component :is="h(Td, { column, record, index }, $slots)" />
    </template>
  </div>
</template>

<script lang="ts" setup>
  import { watchImmediate, wheneverEffectScopeImmediate } from '@mixte/use';
  import { useElementSize } from '@vueuse/core';
  import { h, onMounted, ref } from 'vue';
  import { useShared } from '../composables/useShared';
  import { useVirtual } from '../composables/useVirtual';
  import Td from './td.vue';

  interface Props {
    record: Record<string, any>;
    index: number;
  }

  const props = defineProps<Props>();

  const trRef = ref<HTMLDivElement>();

  const { props: tableProps } = useShared()!;

  onMounted(() => {
    wheneverEffectScopeImmediate(() => tableProps.virtual, () => {
      const { updateRowHeight } = useVirtual()!;
      const height = useElementSize(trRef).height;

      watchImmediate(height, () => {
        updateRowHeight(props.index, trRef.value?.clientHeight ?? 0);
      });
    });
  });
</script>
