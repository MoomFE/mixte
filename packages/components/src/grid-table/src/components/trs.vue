<!-- 所有表体表格行 -->

<template>
  <template v-if="props.virtual">
    <!-- 虚拟表格占位 -->
    <div
      v-if="isLegacyRenderMode"
      class="mixte-gt-virtual-placeholder"
      :style="{ gridColumnStart: `span ${columns?.length}` }"
    />
    <!-- 虚拟表格 -->
    <component
      v-for="(node, index) in data" :key="node.key"
      :is="h(Tr, { node, index: visibleStart + index }, $slots)"
    />
  </template>
  <template v-else>
    <component
      v-for="(node, index) in displayedData" :key="node.key"
      :is="h(Tr, { node, index }, $slots)"
    />
  </template>
</template>

<script lang="ts" setup>
  import type { GridTableFieldsSlots } from '@mixte/components/grid-table/types';
  import { h } from 'vue';
  import { useShared } from '../composables/useShared';
  import { useTreeData } from '../composables/useTreeData';
  import { useVirtual } from '../composables/useVirtual';
  import Tr from './tr.vue';

  defineSlots<GridTableFieldsSlots<any>>();

  const { props, columns, isLegacyRenderMode } = useShared()!;
  const { displayedData } = useTreeData()!;
  const { data, visibleStart } = useVirtual()!;
</script>
