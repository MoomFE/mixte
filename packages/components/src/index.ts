export * from './auto-grid/index';
