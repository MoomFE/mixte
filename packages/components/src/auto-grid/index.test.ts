import { MixteAutoGrid } from '@mixte/components';
import { mount } from '@vue/test-utils';

describe('AutoGrid', () => {
  const emptyOptions = {
    template: '<MixteAutoGrid />',
    components: {
      MixteAutoGrid,
    },
  };

  test('未给组件传入子节点', () => {
    const wrapper = mount(emptyOptions);
    expect(wrapper.html()).matchSnapshot();
  });

  test('给组件传入各个类型的子节点', () => {
    const wrapper = mount({
      ...emptyOptions,
      template: `
        <MixteAutoGrid>

          123

          <!-- comment -->

          <div>456</div>

          <MixteAutoGrid />

          <template v-for="i in 6" :key="i">
            <div>item-{{ i }}</div>
          </template>

          <div v-if="false">789</div>

        </MixteAutoGrid>
      `,
    });

    expect(wrapper.html()).matchSnapshot();
  });
});
