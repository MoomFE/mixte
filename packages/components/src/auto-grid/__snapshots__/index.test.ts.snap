// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AutoGrid > 未给组件传入子节点 1`] = `"<div style=\\"display: grid; grid-template-columns: repeat(NaN, 1fr); column-gap: 0px; row-gap: 0px;\\"></div>"`;

exports[`AutoGrid > 给组件传入各个类型的子节点 1`] = `
"<div style=\\"display: grid; grid-template-columns: repeat(NaN, 1fr); column-gap: 0px; row-gap: 0px;\\">
  <div style=\\"overflow: hidden;\\"> 123 </div>
  <div style=\\"overflow: hidden;\\">
    <div>456</div>
  </div>
  <div style=\\"overflow: hidden;\\">
    <div style=\\"display: grid; grid-template-columns: repeat(NaN, 1fr); column-gap: 0px; row-gap: 0px;\\"></div>
  </div>
  <div style=\\"overflow: hidden;\\">
    <div>item-1</div>
  </div>
  <div style=\\"overflow: hidden;\\">
    <div>item-2</div>
  </div>
  <div style=\\"overflow: hidden;\\">
    <div>item-3</div>
  </div>
  <div style=\\"overflow: hidden;\\">
    <div>item-4</div>
  </div>
  <div style=\\"overflow: hidden;\\">
    <div>item-5</div>
  </div>
  <div style=\\"overflow: hidden;\\">
    <div>item-6</div>
  </div>
</div>"
`;
